### ISSUE TITLE
Signature vs body folding

### ISSUE THREAD

**Issue by pedrorrivero** [microsoft/pylance-release#3385]

**Comment by debonte** [comment]
We previously discussed this in https://github.com/microsoft/pylance-release/issues/2269. Take a look at the three options I described in [this comment](https://github.com/microsoft/pylance-release/issues/2269#issuecomment-1018064327). 

Option 1 is your preferred behavior.
Option 2 is the current behavior.
Option 3 is an approach that I think may make everyone happy.

Let me know what you think.

**Comment by pedrorrivero** [comment]
Thanks for the quick reply @debonte ! I agree with you, option 3 is the best approach! Thanks for putting all this together 😉

**Comment by debonte** [comment]
> Only concern with option 3 really is that it would mess up the levels so that shortcuts will not be consistent (e.g. folding level 2 would mean folding the body in some functions, and folding something internal in other cases depending on whether the signature is multiline or not)

I'm not sure I understand.

To recap, option 3 means that in the following screenshot, folding line 22 collapses the body (leaving the parameter list visible) and folding line 18 collapses the entire function definition (arguments and body):
![image](https://user-images.githubusercontent.com/12820898/197071583-6e18f7a1-5c97-4155-8157-d0a57dd141b7.png)

I think your concern is that with single-line function headers, users might think that the next `>` icon below the `def` will collapse the function body? In the example below, folding line 189 collapses the entire function. If we go with option 3, are you concerned that users will think that folding line 190 would collapse the body of the function, simply because it's the next folding icon below the `def` line?
![image](https://user-images.githubusercontent.com/12820898/197071919-44ca61c1-f05a-4f41-91e2-9a943dc7b823.png)

If that's your concern, I'm not sure how to address it. That would have been a problem with the old (2021, aka option 1) behavior as well. Also, this would seem like an argument for sticking with the current behavior -- a single, consistent folding location for functions.

If I'm misunderstanding your point, could you show me some code snippets to explain what you're describing?

**Comment by debonte** [comment]
Some other thoughts:
1. What if we changed the text of the folding line when folding functions with multiline signatures? So if you folded line 18 of `my_func` above, and it folded the entire function, the text shown on line 18 would be `def my_func(a, b, c)` instead of `def my_func(`. This would require vscode to support LSP's `FoldingRange.collapsedText` field. See https://github.com/microsoft/vscode/issues/70794.
2. We could have a config setting to switch between 2+ function folding behaviors.

**Comment by pedrorrivero** [comment]
This is what I mean. Take the following code snippet as an example, folding points are labeled alphabetically:
```python
def single_line():          # A
    if 0 == 1:              # B
        pass

def multi_line(             # C
    *args,
):                          # D
    pass
```

Folding level 1 means:
- Under option 1 → `#A`, `#C` and `#D` (exposed)
- Under option 3 → `#A` and `#C` (`#D` is hidden)

Folding level 2 means:
- Under option 1 →  `#B`
- Under option 3 → `#B` and `#D`

Notice how folding level 1 is roughly equivalent and equivalent; with the added choice of displaying the signature under option 1 afterwards.

However, folding level 2 under option 3 messes the structure of the document, since one function will get its body folded and the other only an internal `if` statement. And the situation gets worse if you start considering these as methods within a class (which, if the class is large, I like to fold automatically with a key binding for quickly looking through it).

You can toy around with this idea and notice that the real issue with option 3 is that it assigns different precedence levels to things which are structurally at the same "height": throwing the body under the signature instead of keeping them aligned in the hierarchy.

---

> We could have a config setting to switch between 2+ function folding behaviors

This would be a nice way to end the controversy, let users choose what they want.

> What if we changed the text of the folding line when folding functions with multiline signatures?

In which case this would help remind them what their selected choice is 🙂  but it is not a huge concern of mine. I think people can get a grasp of the behavior by simply clicking a few times

**Comment by debonte** [comment]
Thanks. I'm embarrassed to say that I wasn't familiar with the `Fold Level #` commands until now.

I'll need to think about this some more and play with the different behaviors. Thanks for educating me. :)

**Comment by debonte** [comment]
For future reference, here are some screenshots I captured using the three options on the following code:

```python
def single_line():          # A
    if 0 == 1:              # B
        pass

def multi_line(             # C
    *args,
):                          # D
    if 0 == 1:
        pass

class Class:
    def single_line(self):  # A
        if 0 == 1:          # B
            pass

    def multi_line(         # C
        self,           
        *args,
    ):                      # D
        pass

def another():
    pass
```

Folding level 1 (option 1 on left, option 2 center, option 3 on right):
![FoldingLevel1](https://user-images.githubusercontent.com/12820898/213793642-2808f1b4-b224-4947-9295-7e103b3c75e3.png)

Folding level 2:
![FoldingLevel2](https://user-images.githubusercontent.com/12820898/213793663-9950d9b9-5f76-4671-a01e-b1943b806468.png)

**Comment by debonte** [comment]
I agree that option 3's behavior with the `Fold Level #` commands is weird and I think it rules that out as an option.

If there's sufficient interest, we can add an option to choose between options 1 and 2 for folding functions (and classes) with multi-line signatures/headers. However, given that no one else has upvoted or commented on this issue, it's not clear that it's broadly needed.

@pedrorrivero, can you try this setting to see if it is acceptable to you as a workaround in the meantime?

```json
"[python]": {
    "editor.foldingStrategy": "indentation",
},
```

**Comment by debonte** [comment]
@pedrorrivero, ping? See my comment above.

**Comment by pedrorrivero** [comment]
Apologies, I've had very busy weeks. Let me take a look at this over the weekend! Thanks for pinging 🙂

**Comment by pedrorrivero** [comment]
This looks very nice indeed, for what I have been able to see it does exactly what we discussed for option 1, thanks a lot for taking a look and providing a solution.

Looking forward to that extra option in the future if there is interest 😄 Let me know if I can help in any other way

**Comment by pedrorrivero** [comment]
This looks very nice indeed, for what I have been able to see it does exactly what we discussed for option 1, thanks a lot for taking a look and providing a solution @debonte.

Looking forward to that extra option in the future if there is interest 😄 Let me know if I can help in any other way
